lua/ttt2/libraries/migrations.lua

```diff
@@ -0,0 +1,184 @@
+		return false
+	end
+
+	migrations.cachedCommands[#migrations.cachedCommands + 1] = command
+
+	return true
+end
+
+---
+-- Runs all missing forward gamemode migrations.
+-- @return boolean `true` if the desired version was successfully migrated and `false` in case of an error.
+-- @realm shared
+function migrations.Apply()
+	if not migrations.isLoaded then
+		migrations.Load()
+	end
+
+	local migrationSuccess = true
+	local errorMessage = ""
+	local databaseORM = migrations.GetORM()
+
+	for fileName, commands in SortedPairs(migrations.commands) do
+		local fileInfo = databaseORM:Find(fileName)
+
+		-- As ORM currently gives out strings check it
+		if fileInfo and (fileInfo.wasSuccess == true or fileInfo.wasSuccess == "true") then continue end
+
+		MsgN("[TTT2] Migrating: ", fileName)
+
+		for j = 1, #commands do
+			migrationSuccess, errorMessage = pcall(commands[j].Upgrade, commands[j])
+
+			if not migrationSuccess then break end
+		end
+
+		local date = os.date("%Y/%m/%d - %H:%M:%S", os.time())
+
+		if fileInfo then
+			fileInfo.date = date
+			fileInfo.wasSuccess = migrationSuccess
+			fileInfo.wasUpgrade = true
+			fileInfo:Save()
+		else
+			databaseORM:New({
+				name = fileName,
+				date = date,
+				wasSuccess = migrationSuccess,
+				wasUpgrade = true
+			}):Save()
+		end
+
+		if not migrationSuccess then break end
+	end
+
+	if not migrationSuccess then
+		ErrorNoHalt("[TTT2] Migration failed. Error:\n" .. tostring(errorMessage) .. "\n")
+	else
+		MsgN("[TTT2] Migration was successful.")
+	end
+
+	return migrationSuccess
+end
+
+---
+-- Callback function that combines commands with fileNames
+-- @realm shared
+local function fileIncludedCallback(filePath, folderPath)
+	fileName = string.TrimLeft(filePath, folderPath)
+
+	migrations.commands[fileName] = migrations.cachedCommands
+
+	-- Empty cache for next file
+	migrations.cachedCommands = {}
+end
+
+---
+-- Loads migration-files and accepts added commands per file
+-- @realm shared
+function migrations.Load()
+	MsgN("[TTT2] Loading Migrations ...")
+	migrations.cachedCommands = {}
+
+	if CLIENT then
+		fileloader.LoadFolder(migrations.folderPath .. "client/", false, CLIENT_FILE, function(filePath, folderPath)
+			MsgN("Added TTT2 client migration file: ", filePath)
+			fileIncludedCallback(filePath, folderPath)
+		end)
+	end
+
+	if SERVER then
+		fileloader.LoadFolder(migrations.folderPath .. "client/", false, CLIENT_FILE, function(filePath, folderPath)
+			MsgN("Marked TTT2 client migration file for distribution: ", filePath)
+		end)
+
+		fileloader.LoadFolder(migrations.folderPath .. "server/", false, SERVER_FILE, function(filePath, folderPath)
+			MsgN("Added TTT2 server migration file: ", filePath)
+			fileIncludedCallback(filePath, folderPath)
+		end)
+	end
+
+	fileloader.LoadFolder(migrations.folderPath .. "shared/", false, SHARED_FILE, function(filePath, folderPath)
+		MsgN("Added TTT2 shared migration file: ", filePath)
+		fileIncludedCallback(filePath, folderPath)
+	end)
```

Comment:
Just wondering. What would be a usecase for shared migrations?

Reply:
Same ConvarNames or database entries. Not sure if thats important though. Just wanted to be inclusive for the general name schemes.

Reply:
What about just making them shared and have the developer take the decision as usual via ifServer then stuff? Would be much simpler

Reply:
Would be fine by me.